buffer/src/main/java/io/netty/buffer/TestingPooledByteBufAllocator.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.buffer;
+
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakTracker;
+
+import java.io.Closeable;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A {@link ByteBufAllocator} for unit testing.  This allocator performs leak tracking
+ * on all allocated buffers.  At the end the test, call {@link #close} to assert
+ * that all buffers have been freed.
+ */
+public class TestingPooledByteBufAllocator extends PooledByteBufAllocator implements Closeable {
```

Comment:
I wonder if this should better go into src/main/test and we should publish a test-jar to depend on ? 

@trustin @ejona86 @carl-mastrangelo  thoughts ?

Reply:
gRPC uses a separate artifact (grpc-testing), which has been working well. But we also are able to make interesting dependency ordering, like grpc-testing depends on grpc-core, and grpc-core's tests depend on grpc-testing. That doesn't seem like a problem for this one class though.

Using a separate classifier in netty-buffer seems fine in many ways, although it probably won't show up in javadocs which would be unfortunate. With a separate testing artifact, you have to chose how many dependencies you're willing to have, so maybe the classifier is lighter-weight since it's not a big deal to have a separate jar for each existing artifact.

Reply:
I don't think putting this in test-jar would work since you end up pulling in testing deps like Junit.  We could move this to a new module, but we would need to switch the package due to the Java 9 split-module rule, and this would require making the new methods on `AbstractByteBufAllocator` protected (instead of package protected).

Reply:
+1 to a test jar.  There are some  testing utility functions / class I would like to add to Netty, but don't have anywhere to put them.